### PR TITLE
Multiple ports support

### PR DESCRIPTION
**Pull request by shamsimam**

### CODE REVIEW

**Comment by shamsimam** [comment]
Summary:

- Allows a service to request multiple ports from Waiter, currently, maximum of 10 is allowed.
- Waiter expects the service will use the first port for the web server.
- `ServiceInstance` reflects this choice by having two fields:
-- `port`: the port for the web server,
-- `auxiliary-ports`: any additional ports reserved for the service.

File: waiter/src/waiter/marathon.clj [hunk 1]
```diff
@@ -181,24 +183,24 @@
...
-        {:strs [grace-period-secs run-as-user]} service-description
+        {:strs [cmd cpus disk grace-period-secs mem ports restart-backoff-factor run-as-user]} service-description
         home-path (str home-path-prefix run-as-user)]
     {:id service-id
      :env (scheduler/environment service-id service-description service-id->password-fn home-path)
      :user run-as-user
-     :cmd (get service-description "cmd")
-     :disk (get service-description "disk")
-     :mem (get service-description "mem")
-     :ports (get service-description "ports" [0])
```
Comment (dposada) [inline waiter/src/waiter/marathon.clj]: I understand what your new code does, but I'm not sure I understand how the old code worked. What if I provided a number like 4 for "ports"?

File: waiter/src/waiter/schema.clj [hunk 2]
```diff
@@ -45,6 +45,10 @@
                                 (instance? Pattern (first %))
                                 (string? (second %))))])
 
+(def valid-ports
```
Comment (dposada) [inline waiter/src/waiter/schema.clj]: Maybe `valid-number-of-ports` or `valid-port-count`?

File: waiter/test/waiter/marathon_test.clj [hunk 3]
```diff
@@ -102,7 +102,7 @@
                                                                      :started-at "2014-09-13T002456.965Z"}),
                                                                   (scheduler/make-ServiceInstance
                                                                     {:id "test-app-1234.C", :service-id "test-app-1234", :healthy? false,
-                                                                     :host "10.141.141.13", :port 41234,
+                                                                     :host "10.141.141.13", :port 41234, :auxiliary-ports [12321 90384 56463],
```
Comment (dposada) [inline waiter/test/waiter/marathon_test.clj]: Nitpick: I would favor one entry per line

Comment (shamsimam) [inline waiter/test/waiter/marathon_test.clj]: Done.

File: waiter/src/waiter/marathon.clj [hunk 4]
```diff
@@ -130,7 +130,9 @@
                                   {:id instance-id
                                    :started-at (str (:startedAt %))
                                    :healthy? (healthy?-fn %)
-                                   :port (first (:ports %))})))
+                                   ;; first port must be used for the web server, auxiliary ports can be used freely.
+                                   :port (-> % :ports first)
+                                   :auxiliary-ports (-> % :ports rest vec)})))
```
Comment (sradack) [inline waiter/src/waiter/marathon.clj]: Something like `extra-ports` would be easier to say and spell.

File: waiter/src/waiter/shell_scheduler.clj [hunk 5]
```diff
@@ -128,10 +128,11 @@
 
 (defn launch-instance
   "Launches a new process for the given service-id"
-  [service-id working-dir-base-path command environment port->reservation-atom port-range]
+  [service-id working-dir-base-path command environment num-ports port->reservation-atom port-range]
   (when-not command
     (throw (ex-info "The command to run was not supplied" {:service-id service-id})))
-  (let [port (reserve-port! port->reservation-atom port-range)
+  (let [reserved-ports (repeatedly num-ports #(reserve-port! port->reservation-atom port-range))
```
Comment (sradack) [inline waiter/src/waiter/shell_scheduler.clj]: The port count should be passed to `reserve-port!`.  What happens if you run out of ports?

**Comment by shamsimam** [comment]
@dposada any further feedback?

File: waiter/src/waiter/shell_scheduler.clj (hunk 5 above)
Comment (shamsimam) [inline waiter/src/waiter/shell_scheduler.clj]: Done, ended up using `atom`.

The following did not compile:
```
(defn reserve-ports!
  "Reserves num-ports available ports on the host, from the (optionally) provided range"
  [num-ports port->reservation-atom port-range]
  (try
    (loop [reserved-ports []
           remaining num-ports]
      (if (pos? remaining)
        (let [reserved-port-or-exception
              (try
                (reserve-port! port->reservation-atom port-range)
                (catch Throwable ex
                  (ex-info (str "Unable to reserve " num-ports " ports.")
                           {:num-successfully-reserved-before-error (count reserved-ports)
                            :reserved-ports reserved-ports}
                           ex)))]
          (if (instance? Throwable reserved-port-or-exception)
            reserved-port-or-exception
            (recur (conj reserved-ports reserved-port-or-exception) (dec remaining)))))
      reserved-ports)
    (catch Exception ex
      (throw ex))))
```

Error message:
```      
CompilerException java.lang.UnsupportedOperationException: Can only recur from tail position, compiling:(/Users/shamsimam/projects/github-projects/twosigma-waiter/waiter/src/waiter/shell_scheduler.clj:146:13) 
```

File: waiter/src/waiter/marathon.clj (hunk 4 above)
Comment (shamsimam) [inline waiter/src/waiter/marathon.clj]: Renamed.

**Comment by shamsimam** [comment]
@sradack addressed your feedback.

**Comment by dposada** [comment]
A couple of comments:

- Can we document the rationale for multiple port support? In other words, why would someone want to do this? What's the use case?
- Should we add an integration test that confirms multiple ports become available to the backend?

**Comment by shamsimam** [comment]
@dposada 
Meant to respond earlier, but then got side-tracked and forgot, sorry.

Is the parameter documentation not sufficient?

Also, there is an integration test (`test-multiple-ports`) that validates that an instance has been allocated ports.

**Comment by shamsimam** [comment]
@dposada improved the integration test and the documentation :)

**Comment by shamsimam** [comment]
Passing build: https://travis-ci.org/twosigma/waiter/builds/248603064

File: waiter/docs/service-checklist.md [hunk 6]
```diff
@@ -14,6 +14,14 @@ myservice/bin/run-my-service -p $PORT0
 
 Take care to escape variables that start with `$` such that they are not interpreted by your local bash shell.
 
+## Additional ports
+
+Each service is assigned a single port by default which must be used by your HTTP service to listen to.
```
Comment (dposada) [inline waiter/docs/service-checklist.md]: "to listen on"?

Comment (shamsimam) [inline waiter/docs/service-checklist.md]: Done.

**Comment by shamsimam** [comment]
Passing build: https://travis-ci.org/twosigma/waiter/builds/248630083

**Pull request merged**